### PR TITLE
HEVC Direct Play override for Profile Version

### LINKED ISSUE #724: HEVC Content Being Transcoded on app version 1.5 because of VideoProfile
**BUG**
Since upgrading to app version 1.5, all HEVC files are being transcoded to h264.  This also seems to be triggering an audio transcode as well, but the root issue is the video
Jellyfin 1-5
Jellyfin Dev
.
The screenshots show the same file, in this case my Bluray rip of The Batman.  It is encoded in HEVC/EAC3.  WIth 1.5 it transcodes.  On the dev channel version I have installed from a week or so ago, the file direct plays without issue.

**Expected Behavior**
HEVC content should Direct Play as it did on previous versions of the Jellyfin Roku app.

**JELLYFIN LOG ENTRY:**
`2022-07-31 01:09:15.715 +00:00] [INF] [53] Jellyfin.Api.Helpers.MediaInfoHelper: Profile: "VideoCodecProfile", DirectPlay=false. Reason="Unknown Profile".VideoProfile Condition: EqualsAny. ConditionValue: "main". IsRequired: False. Path: "nfs:/REDACTED/The ``Batman.mkv"`

**FULL TRANSCODING LOG**
```
`/videos/74d3ffb1-4242-4129-b5ee-14eeb21e3a02/hls1/main/1491.ts

{"Protocol":0,"Id":"74d3ffb142424129b5ee14eeb21e3a02","Path":"/media/Linux_HD_4TB2/Video/1 - Movies/Batman Movies/The Batman (2022)/The Batman.mkv","EncoderPath":null,"EncoderProtocol":null,"Type":0,"Container":"mkv,webm","Size":10047764480,"Name":"The Batman","IsRemote":false,"ETag":"b136d8847ffdbfb49b0d6e0943e611b7","RunTimeTicks":105715662848,"ReadAtNativeFramerate":false,"IgnoreDts":false,"IgnoreIndex":false,"GenPtsInput":false,"SupportsTranscoding":true,"SupportsDirectStream":true,"SupportsDirectPlay":true,"IsInfiniteStream":false,"RequiresOpening":false,"OpenToken":null,"RequiresClosing":false,"LiveStreamId":null,"BufferMs":null,"RequiresLooping":false,"SupportsProbing":true,"VideoType":0,"IsoType":null,"Video3DFormat":null,"MediaStreams":[{"Codec":"hevc","CodecTag":null,"Language":null,"ColorRange":null,"ColorSpace":"bt709","ColorTransfer":"bt709","ColorPrimaries":"bt709","Comment":null,"TimeBase":"1/1000","CodecTimeBase":"1001/24000","Title":null,"VideoRange":"SDR","localizedUndefined":null,"localizedDefault":null,"localizedForced":null,"DisplayTitle":"1080p HEVC SDR","NalLengthSize":null,"IsInterlaced":false,"IsAVC":null,"ChannelLayout":null,"BitRate":7603614,"BitDepth":null,"RefFrames":1,"PacketLength":null,"Channels":null,"SampleRate":null,"IsDefault":true,"IsForced":false,"Height":804,"Width":1920,"AverageFrameRate":23.976025,"RealFrameRate":23.976025,"Profile":"Main 10","Type":1,"AspectRatio":"2.40:1","Index":0,"Score":null,"IsExternal":false,"DeliveryMethod":null,"DeliveryUrl":null,"IsExternalUrl":null,"IsTextSubtitleStream":false,"SupportsExternalStream":false,"Path":null,"PixelFormat":"yuv420p10le","Level":120,"IsAnamorphic":null},{"Codec":"eac3","CodecTag":null,"Language":"eng","ColorRange":null,"ColorSpace":null,"ColorTransfer":null,"ColorPrimaries":null,"Comment":null,"TimeBase":"1/1000","CodecTimeBase":"1/48000","Title":"Dolby Digital\u002B 5.1","VideoRange":null,"localizedUndefined":null,"localizedDefault":null,"localizedForced":null,"DisplayTitle":"Dolby Digital\u002B 5.1 - Eng - 6 ch - Default","NalLengthSize":null,"IsInterlaced":false,"IsAVC":null,"ChannelLayout":null,"BitRate":640000,"BitDepth":null,"RefFrames":null,"PacketLength":null,"Channels":6,"SampleRate":48000,"IsDefault":true,"IsForced":false,"Height":null,"Width":null,"AverageFrameRate":null,"RealFrameRate":null,"Profile":null,"Type":0,"AspectRatio":null,"Index":1,"Score":null,"IsExternal":false,"DeliveryMethod":null,"DeliveryUrl":null,"IsExternalUrl":null,"IsTextSubtitleStream":false,"SupportsExternalStream":false,"Path":null,"PixelFormat":null,"Level":0,"IsAnamorphic":null},{"Codec":"ac3","CodecTag":null,"Language":"eng","ColorRange":null,"ColorSpace":null,"ColorTransfer":null,"ColorPrimaries":null,"Comment":null,"TimeBase":"1/1000","CodecTimeBase":"1/48000","Title":"Dolby Digital 5.1","VideoRange":null,"localizedUndefined":null,"localizedDefault":null,"localizedForced":null,"DisplayTitle":"Dolby Digital 5.1 - Eng","NalLengthSize":null,"IsInterlaced":false,"IsAVC":null,"ChannelLayout":"5.1","BitRate":640000,"BitDepth":null,"RefFrames":null,"PacketLength":null,"Channels":6,"SampleRate":48000,"IsDefault":false,"IsForced":false,"Height":null,"Width":null,"AverageFrameRate":null,"RealFrameRate":null,"Profile":null,"Type":0,"AspectRatio":null,"Index":2,"Score":null,"IsExternal":false,"DeliveryMethod":null,"DeliveryUrl":null,"IsExternalUrl":null,"IsTextSubtitleStream":false,"SupportsExternalStream":false,"Path":null,"PixelFormat":null,"Level":0,"IsAnamorphic":null},{"Codec":"srt","CodecTag":null,"Language":"eng","ColorRange":null,"ColorSpace":null,"ColorTransfer":null,"ColorPrimaries":null,"Comment":null,"TimeBase":null,"CodecTimeBase":null,"Title":null,"VideoRange":null,"localizedUndefined":"Undefined","localizedDefault":"Default","localizedForced":"Forced","DisplayTitle":"Eng","NalLengthSize":null,"IsInterlaced":false,"IsAVC":null,"ChannelLayout":null,"BitRate":null,"BitDepth":null,"RefFrames":null,"PacketLength":null,"Channels":null,"SampleRate":null,"IsDefault":false,"IsForced":false,"Height":null,"Width":null,"AverageFrameRate":null,"RealFrameRate":null,"Profile":null,"Type":2,"AspectRatio":null,"Index":3,"Score":null,"IsExternal":true,"DeliveryMethod":null,"DeliveryUrl":null,"IsExternalUrl":null,"IsTextSubtitleStream":true,"SupportsExternalStream":true,"Path":"/media/Linux_HD_4TB2/Video/1 - Movies/Batman Movies/The Batman (2022)/The Batman.en.srt","PixelFormat":null,"Level":null,"IsAnamorphic":null}],"MediaAttachments":[],"Formats":[],"Bitrate":8883614,"Timestamp":null,"RequiredHttpHeaders":{},"TranscodingUrl":null,"TranscodingSubProtocol":null,"TranscodingContainer":null,"AnalyzeDurationMs":null,"DefaultAudioStreamIndex":null,"DefaultSubtitleStreamIndex":null}

/usr/lib/jellyfin-ffmpeg/ffmpeg -ss 01:14:33.000 -hwaccel vaapi -noaccurate_seek -hwaccel_output_format vaapi -vaapi_device /dev/dri/renderD128 -i file:"/media/Linux_HD_4TB2/Video/1 - Movies/Batman Movies/The Batman (2022)/The Batman.mkv" -map_metadata -1 -map_chapters -1 -threads 0 -map 0:0 -map 0:1 -map -0:s -codec:v:0 h264_vaapi -b:v 12672690 -maxrate 12672690 -bufsize 25345380 -profile:v:0 high -level 41  -force_key_frames:0 "expr:gte(t,4473+n_forced*3)" -vf "format=nv12|vaapi,hwupload,scale_vaapi=format=nv12" -start_at_zero -vsync -1 -codec:a:0 aac -ac 6 -ab 640000 -copyts -avoid_negative_ts disabled -max_muxing_queue_size 2048 -f hls -max_delay 5000000 -hls_time 3 -hls_segment_type mpegts -start_number 1491 -hls_segment_filename "/config/transcodes/f57441f799e51fd3ef73299d1bff61c4%d.ts" -hls_playlist_type vod -hls_list_size 0 -y "/config/transcodes/f57441f799e51fd3ef73299d1bff61c4.m3u8"


ffmpeg version 4.3.2-Jellyfin Copyright (c) 2000-2021 the FFmpeg developers
  built with gcc 8 (Debian 8.3.0-6)
  configuration: --prefix=/usr/lib/jellyfin-ffmpeg --target-os=linux --extra-version=Jellyfin --disable-doc --disable-ffplay --disable-shared --disable-libxcb --disable-sdl2 --disable-xlib --enable-gpl --enable-version3 --enable-static --enable-libfontconfig --enable-fontconfig --enable-gmp --enable-gnutls --enable-libass --enable-libbluray --enable-libdrm --enable-libfreetype --enable-libfribidi --enable-libmp3lame --enable-libopus --enable-libtheora --enable-libvorbis --enable-libdav1d --enable-libwebp --enable-libvpx --enable-libx264 --enable-libx265 --enable-libzvbi --enable-libzimg --arch=amd64 --enable-opencl --enable-vaapi --enable-amf --enable-libmfx --enable-vdpau --enable-cuda --enable-cuda-llvm --enable-cuvid --enable-nvenc --enable-nvdec --enable-ffnvcodec
  libavutil      56. 51.100 / 56. 51.100
  libavcodec     58. 91.100 / 58. 91.100
  libavformat    58. 45.100 / 58. 45.100
  libavdevice    58. 10.100 / 58. 10.100
  libavfilter     7. 85.100 /  7. 85.100
  libswscale      5.  7.100 /  5.  7.100
  libswresample   3.  7.100 /  3.  7.100
  libpostproc    55.  7.100 / 55.  7.100
Guessed Channel Layout for Input Stream #0.1 : 5.1
Input #0, matroska,webm, from 'file:/REDACTED/The Batman.mkv':
  Metadata:
    title           : The Batman
    creation_time   : 2022-07-07T16:15:35.000000Z
    ENCODER         : Lavf58.76.100
  Duration: 02:56:11.57, start: 0.000000, bitrate: 7603 kb/s
    Chapter #0:0: start 0.000000, end 721.221000
    Metadata:
      title           : Chapter 1
    Chapter #0:1: start 721.221000, end 1401.734000
    Metadata:
      title           : Chapter 2
    Chapter #0:2: start 1401.734000, end 1914.079000
    Metadata:
      title           : Chapter 3
    Chapter #0:3: start 1914.079000, end 2504.502000
    Metadata:
      title           : Chapter 4
    Chapter #0:4: start 2504.502000, end 3096.677000
    Metadata:
      title           : Chapter 5
    Chapter #0:5: start 3096.677000, end 3612.901000
    Metadata:
      title           : Chapter 6
    Chapter #0:6: start 3612.901000, end 4232.395000
    Metadata:
      title           : Chapter 7
    Chapter #0:7: start 4232.395000, end 4715.127000
    Metadata:
      title           : Chapter 8
    Chapter #0:8: start 4715.127000, end 5354.558000
    Metadata:
      title           : Chapter 9
    Chapter #0:9: start 5354.558000, end 6153.064000
    Metadata:
      title           : Chapter 10
    Chapter #0:10: start 6153.064000, end 6736.897000
    Metadata:
      title           : Chapter 11
    Chapter #0:11: start 6736.897000, end 7101.011000
    Metadata:
      title           : Chapter 12
    Chapter #0:12: start 7101.011000, end 7779.689000
    Metadata:
      title           : Chapter 13
    Chapter #0:13: start 7779.689000, end 8474.383000
    Metadata:
      title           : Chapter 14
    Chapter #0:14: start 8474.383000, end 8851.343000
    Metadata:
      title           : Chapter 15
    Chapter #0:15: start 8851.343000, end 9516.090000
    Metadata:
      title           : Chapter 16
    Chapter #0:16: start 9516.090000, end 10018.759000
    Metadata:
      title           : Chapter 17
    Chapter #0:17: start 10018.759000, end 10571.520000
    Metadata:
      title           : Chapter 18
    Stream #0:0: Video: hevc (Main 10), yuv420p10le(tv, bt709), 1920x804 [SAR 1:1 DAR 160:67], 23.98 fps, 23.98 tbr, 1k tbn, 23.98 tbc (default)
    Metadata:
      DURATION        : 02:56:11.566000000
    Stream #0:1(eng): Audio: eac3, 48000 Hz, 5.1, fltp (default)
    Metadata:
      title           : Dolby Digital+ 5.1
      DURATION        : 02:56:11.552000000
    Stream #0:2(eng): Audio: ac3, 48000 Hz, 5.1(side), fltp, 640 kb/s
    Metadata:
      title           : Dolby Digital 5.1
      DURATION        : 02:56:11.552000000
Stream mapping:
  Stream #0:0 -> #0:0 (hevc (native) -> h264 (h264_vaapi))
  Stream #0:1 -> #0:1 (eac3 (native) -> aac (native))
```

### PR DESCRIPTION
**Pull request by neilsb**
Add user profile option to attempt direct play of HEVC content where the profile lever of the media is not supported by the Roku device.  If it fails, silently fall back to transcoding.   

The functionality exists for h264 media, this is just adding an option for HEVC content.

**Changes**
 - New User Setting to attempt direct play
 - Translation for the new HEVC setting and h264 setting that was missed

**Issues**
fixes #724

### CODE REVIEW

**Comment by neilsb** [comment]
Not overly keen of the way the `if` has been structured in VidepPlayer.brs.  But since Roku doesn't support multi line if statements, and the line was so long, thought this made it slightly more readable.

File: locale/en_US/translations.ts [hunk 1]
```diff
@@ -692,5 +692,25 @@
...
+            <translation>Attempt H.264 Direct Play (Profile Lvl)</translation>
+            <extracomment>Settings Menu - Title for option</extracomment>
+        </message>
+        <message>
+            <source>Attempt Direct Play for H.264 media with unsupported profile levels (> 4.2) before falling back to transcoding if it fails.</source>
+            <translation>Attempt Direct Play for H.264 media with unsupported profile levels (> 4.2) before falling back to transcoding if it fails.</translation>
+            <extracomment>Settings Menu - Description for option</extracomment>
+        </message>
+        <message>
+            <source>Attempt HECV Direct Play (Profile Lvl)</source>
```
Comment (crobibero) [inline locale/en_US/translations.ts]:
```suggestion
            <source>Attempt HEVC Direct Play (Profile Lvl)</source>
```

Also line below.

What does "Profile Lvl" mean?

**Comment by neilsb** [comment]
@crobibero Lvl was short for Level. I thought the generally used term was "Profile Level" as opposed to just "Profile" and was trying to save some space when displaying the title. But for the sake of 2 characters, best to stick to "Level"?

**Comment by lakerssuperman** [comment]
I took it for a test drive.  I see in the options that the profile override now exists.  However, when I click on a movie it opens the movie screen with the synopsis and options (play, options, set watched, favorites) and freezes there on every title I click on.  Only pressing the home button on the remote and closing out of the app unfreezes things.  As a result, I wasn't able to test if HEVC video successfully direct plays.

**Comment by neilsb** [comment]
> I took it for a test drive. I see in the options that the profile override now exists. However, when I click on a movie it opens the movie screen with the synopsis and options (play, options, set watched, favorites) and freezes there on every title I click on

@lakerssuperman : It freezes when displaying the movie details?  Before attempting to click _Play_ or anything?   

I'm unable to replicate this, and the code changes shouldn't affect that screen at from my understanding.  Does it happen in all movies or just some?  If it's just some, is there any commonality between them that you can see?

If you turn the setting off in the settings menu does the same happen?

**Comment by lakerssuperman** [comment]
> > I took it for a test drive. I see in the options that the profile override now exists. However, when I click on a movie it opens the movie screen with the synopsis and options (play, options, set watched, favorites) and freezes there on every title I click on
> 
> @lakerssuperman : It freezes when displaying the movie details? Before attempting to click _Play_ or anything?
> 
> I'm unable to replicate this, and the code changes shouldn't affect that screen at from my understanding. Does it happen in all movies or just some? If it's just some, is there any commonality between them that you can see?
> 
> If you turn the setting off in the settings menu does the same happen?

IMG_20220926_163239791

Neil,

Hope all is well.

When I click into a movie it freezes at the screen I attached. I'm unable to select play or use the dpad to select any of the other iptions.  The only thing that works is pressing home to close out of the app. I will double check later, but I believe the HEVC direct play option being toggled had no effect. This was on my 46xx series uktras and as I recall from my tests any item I selected led to the freeze

If you can't replicate it, it could very well be on my end, but as soon as I reverted to the older dev building I've been using everything work correctly again.  I'm happy to do more testing.

**Comment by lakerssuperman** [comment]
> > I took it for a test drive. I see in the options that the profile override now exists. However, when I click on a movie it opens the movie screen with the synopsis and options (play, options, set watched, favorites) and freezes there on every title I click on
> 
> @lakerssuperman : It freezes when displaying the movie details? Before attempting to click _Play_ or anything?
> 
> I'm unable to replicate this, and the code changes shouldn't affect that screen at from my understanding. Does it happen in all movies or just some? If it's just some, is there any commonality between them that you can see?
> 
> If you turn the setting off in the settings menu does the same happen?

I did some further testing and here's what's happening on my end:

1) I tried your HEVC direct play branch on my Roku Ultra 4800x series.  It also freezes.
2) Freezing occurs on the 4800x: as I described above when clicking on movies and TV shows.  TV shows freeze either from navigating to the show via the TV Show library or by clicking on a next up TV episode and using the pop up menu options "start from beginning, go to series, etc".
3) Music playback seems to work correctly.  I was able to navigate to an album, select a song and playback would start as expected.  Individual song playback worked, full album playback worked.  I tried this on several albums.
4) On my 4800x, I'm able to bring up the options menu and scroll to settings.  However, as soon as I click settings the options menu disappears and the app freezes.

**Comment by lakerssuperman** [comment]
Greetings!!

After some help over in the Matrix chat, I now have the correct ropm modules setup and building Unstable works perfectly fine with no hard locks.

Trying to build the HEVC direct play fix repo now with ropm installed gives me the ability to access settings and toggle HEVC directly play, however when attempting to play a video I get a hard lock.  Enabling or disabling HEVC direct play bot h resulted in a hard lock.

I didn't realize I was able to telnet into the Roku and get an error log, but now that I know that this is what my Roku reports: 
https://pastebin.com/wHJ6nSZ5

Thanks!!

**Comment by neilsb** [comment]
@lakerssuperman Thanks -that was useful.  I think it was an error on my part when testing, but pushed an update to the PR which should fix it.  Sorry about that.

**Comment by lakerssuperman** [comment]
> @lakerssuperman Thanks -that was useful. I think it was an error on my part when testing, but pushed an update to the PR which should fix it. Sorry about that.

No need to be sorry.  I'm glad I could help and I appreciate your time in adding this feature.  Thank you!!

**Comment by melat0nin** [comment]
Are there instructions for building and testing this version? I'm keen to give it a go!

**Comment by neilsb** [comment]
> Are there instructions for building and testing this version? I'm keen to give it a go!

Easiest way is to use the output form our CI checks.  Saves having to build anything.

First step is [enable developer mode](https://developer.roku.com/en-gb/docs/developer-program/getting-started/developer-setup.md) on your device if you have not yet done so (Step 1).   This will allow you to sideload apps but will not affect the normal use of your device.  

Then you can access the developer installer (Step 2 from the instructions above) and upload the output form any of our PR Build steps.   

To get the appropriate file to upload: 
 - Goto the PR and at the bottom of the conversation tab you'll see the Checks (Should say "2 successful checks") and just expand the _show all checks_
 - For the `build / run` check, click the Details link
 - Select the `Summary` link at the top left of that page
 - You'll see the "Artifacts" section at the bottom of the page showing the build file you can upload to your Roku (called something similar to  `Jellyfin-Roku-dev-50bc99366de3428d46c8ce0818d943d24eacdf60`

Once uploaded to your Roku, it should appear on the home screen as any normal channel would.  It will always be the last entry on the list, and you can only have one sideloaded app on your device at a time.   You can upload new versions without uninstalling previous ones, as it will just overwrite it.

**Comment by lakerssuperman** [comment]
I just gave the new code a run through and video does play, but I'm getting still getting transcoding even with the HEVC direct play override enabled.

Here's my log from the Roku: https://pastebin.com/3fk19rnF

**Comment by neilsb** [comment]
While playing the content, can you press the up button. This should show media and transcode information, and should display the reason why it's transcoding.

**Comment by lakerssuperman** [comment]
> While playing the content, can you press the up button. This should show media and transcode information, and should display the reason why it's transcoding.

I'm away from my computer right now, but I had the server dashboard open during playback to check that and it reported that the video profile wasn't supported. I will give it a more thorough look when I get home later and see if I can turn anything else up.

**Comment by melat0nin** [comment]
Thanks @neilsb, it installed great. Alas, HEVC main 10 files are still being transcoded for me, despite the setting being enabled (`TranscodeReason=VideoProfileNotSupported`)

Media info from JF web:

> Title1080p HEVC SDR
Codec HEVC
Codec tag hev1
Profile Main 10
Level 120
Resolution 1920x1080
Aspect ratio 16:9
Interlaced No
Framerate 23.976025
Bitrate 2000 kbps
Bit depth 10 bit
Video range SDR
Video range type SDR
Pixel format yuv420p10le
Ref frames 1

**Comment by lakerssuperman** [comment]
> Thanks @neilsb, it installed great. Alas, HEVC main 10 files are still being transcoded for me, despite the setting being enabled (`TranscodeReason=VideoProfileNotSupported`)
> 
> Media info from JF web:
> 
> > Title1080p HEVC SDR
> > Codec HEVC
> > Codec tag hev1
> > Profile Main 10
> > Level 120
> > Resolution 1920x1080
> > Aspect ratio 16:9
> > Interlaced No
> > Framerate 23.976025
> > Bitrate 2000 kbps
> > Bit depth 10 bit
> > Video range SDR
> > Video range type SDR
> > Pixel format yuv420p10le
> > Ref frames 1

Exactly what I'm seeing.  Always glad when it's not just me lol.

**Comment by lakerssuperman** [comment]
@neilsb Hey there, hope all is well.  I did a little more checking with a different clip with a different HEVC profile level and that also failed to direct play, which I figured it would, but just wanted to be sure.  Let me know if you need any more info about the clips I'm using or logs from my Roku.  Thanks again!!

**Comment by neilsb** [comment]
> I did a little more checking with a different clip with a different HEVC profile level and that also failed to direct play

Was this another main-10 clip?  I think that I may have confused things with my terminology and PR wording, which may not be technically correct.  This PR currently only looks at the profile level (4.1, 5.0, 5.1 etc) and not the actual profile (main, main 10) so I think main 10 may we always transcode. 

I think there was another draft PR (#686) that also added support for checking the profiles supported on the device and updating the profile. In think this may be what’s needed to fix main 10 transcoding. 

I’ll check out that and incorporate it in this PR is that helps (and @whiteowl3 is happy with that)

**Comment by lakerssuperman** [comment]
> > I did a little more checking with a different clip with a different HEVC profile level and that also failed to direct play
> 
> Was this another main-10 clip? I think that I may have confused things with my terminology and PR wording, which may not be technically correct. This PR currently only looks at the profile level (4.1, 5.0, 5.1 etc) and not the actual profile (main, main 10) so I think main 10 may we always transcode.
> 
> I think there was another draft PR (#686) that also added support for checking the profiles supported on the device and updating the profile. In think this may be what’s needed to fix main 10 transcoding.
> 
> I’ll check out that and incorporate it in this PR is that helps (and @whiteowl3 is happy with that)

Yes, it was a main 10 clip so that makes sense.

> > I did a little more checking with a different clip with a different HEVC profile level and that also failed to direct play
> 
> Was this another main-10 clip? I think that I may have confused things with my terminology and PR wording, which may not be technically correct. This PR currently only looks at the profile level (4.1, 5.0, 5.1 etc) and not the actual profile (main, main 10) so I think main 10 may we always transcode.
> 
> I think there was another draft PR (#686) that also added support for checking the profiles supported on the device and updating the profile. In think this may be what’s needed to fix main 10 transcoding.
> 
> I’ll check out that and incorporate it in this PR is that helps (and @whiteowl3 is happy with that)

Yes, it was a main 10 clip so that makes sense.

And that would be awesome. Thank you.

**Comment by whiteowl3** [comment]
yeah you are welcome to whatever code in that PR might be helpful.  I wont have time to clean it up for weeks.

**Comment by neilsb** [comment]
I think PR still makes sense to include to allow users to choose to direct play HEVC titles with a higher **profile level** than the box supports.   

Many of the comments in this were due to the video `Profile` rather than the `ProfileLevel`, which is fixed in #686 .  This PR is specifically for `ProfileLevel`

**Comment by kamanwu** [comment]
My Roku TV still cannot play HEVC directly.  Hope this PR will fix the issue.

File: locale/en_US/translations.ts [hunk 2]
```diff
@@ -712,6 +712,25 @@
             <translation>Play Trailer</translation>
         </message>
         <message>
+            <source>Attempt H.264 Direct Play (Profile Lvl)</source>
+            <translation>Attempt H.264 Direct Play (Profile Lvl)</translation>
+            <extracomment>Settings Menu - Title for option</extracomment>
+        </message>
+        <message>
+            <source>Attempt Direct Play for H.264 media with unsupported profile levels (> 4.2) before falling back to transcoding if it fails.</source>
```
Comment (1hitsong) [inline locale/en_US/translations.ts]: If Roku pushes a software change that affect supported profile levels, this text will need to change. Should we simply leave it at "unsupported profile levels?"

File: settings/settings.json [hunk 3]
```diff
@@ -18,12 +18,19 @@
                 "default": "true"
             },
             {
-                "title": "Attempt Direct Play (Profile Lvl)",
+                "title": "Attempt H.264 Direct Play (Profile Lvl)",
```
Comment (1hitsong) [inline settings/settings.json]: Can we we rework the title of the setting to include unsupported profile levels? 

Something such as "Direct Play H.264 Unsupported Profile Levels"

File: settings/settings.json [hunk 4]
```diff
@@ -18,12 +18,19 @@
...
                 "description": "Attempt Direct Play for H.264 media with unsupported profile levels (> 4.2) before falling back to transcoding if it fails.",
                 "settingName": "playback.tryDirect.h264ProfileLevel",
                 "type": "bool",
                 "default": "true"
             },
+            {
+                "title": "Attempt hevc Direct Play (Profile Lvl)",
```
Comment (1hitsong) [inline settings/settings.json]: Same suggestion as above, but HEVC should be all caps.

File: settings/settings.json [hunk 5]
```diff
@@ -18,12 +18,19 @@
...
+                "description": "Attempt Direct Play for hevc media with unsupported profile levels (> 5.1) before falling back to trancoding if it fails.",
```
Comment (1hitsong) [inline settings/settings.json]: Same suggestion as above, but HEVC should be all caps.

**Review by 1hitsong — CHANGES_REQUESTED** [review]

**Review by 1hitsong — APPROVED** [review]

**Pull request merged**